### PR TITLE
update to mxnet==1.0

### PR DESCRIPTION
**Pull request by fhieber**
Update to mxnet==1.0.

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [ ] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./pre-commit.sh` or manual run of pylint & mypy)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by larroy** [comment]
Will you guys run performance regressions after the update?

**Comment by fhieber** [comment]
We will

**Comment by fhieber** [comment]
In training, MXNet 1.0 behaves just like 0.12.1. No differences in speed or memory consumption. Will report on inference later

**Comment by larroy** [comment]
Thanks for the update.
What is the batch size you are using? 

I'm thinking about the issue that we were tracking with @KellenSunderland which can affect GPU usage due to excessive cuda syncthing.

**Comment by fhieber** [comment]
This was a rather small model (1-layer RNN) using a batch size of 80 on 2 Voltas (effective batch size 40 per GPU).

File: sockeye/arguments.py [hunk 1]
```diff
@@ -697,6 +697,15 @@ def add_training_args(params):
...
                                    "training. Default: %(default)s.")
+    train_params.add_argument("--gctype",
+                              type=str,
+                              default=None,
+                              choices=[None, "2bit"],
+                              help='Type of gradient compression to use. Choose "2bit" or None. Default: %(default)s.')
+    train_params.add_argument("--gcthreshold",
+                              type=float,
+                              default=0.5,
+                              help="Threshold for gradient compression if --gctype is '2bit'. Default: %(default)s.")
```
Comment (mjdenkowski) [inline sockeye/arguments.py]: Silly comment: when I see "gc", I think "gradient clipping".  Can we expand this to `--gradient-compression-type` and so on to match `--gradient-clipping-type`?

File: sockeye/inference.py [hunk 2]
```diff
@@ -797,9 +797,7 @@ def _get_inference_input(self, sequences: List[List[str]]) -> Tuple[mx.nd.NDArra
         utils.check_condition(C.PAD_ID == 0, "pad id should be 0")
         source = mx.nd.zeros((len(sequences), bucket_key))
         for j, tokens in enumerate(sequences):
-            ids = data_io.tokens2ids(tokens, self.vocab_source)
-            for i, wid in enumerate(ids):
-                source[j, i] = wid
+            source[j, :len(tokens)] = data_io.tokens2ids(tokens, self.vocab_source)
```
Comment (mjdenkowski) [inline sockeye/inference.py]: Nice that we can simplify like this now!

File: sockeye/arguments.py (hunk 1 above)
Comment (fhieber) [reply sockeye/arguments.py]: good point, makes sense!
I thought of garbage collection :)

**Comment by fhieber** [comment]
CPU-based inference is about 10% faster in MXNet 1.0 w/MKL on a mac , probably due to the performance improvements mentioned in the release notes: https://cwiki.apache.org/confluence/display/MXNET/Apache+MXNet+%28incubating%29+1.0+Release+Notes#ApacheMXNet(incubating)1.0ReleaseNotes-Performance

File: sockeye/inference.py [hunk 3]
```diff
@@ -1042,15 +1033,11 @@ def _beam_search(self,
...
+                (best_hyp_indices[rows], best_word_indices[rows]), \
+                scores_accumulated[rows, 0] = utils.smallest_k(sliced_scores, self.beam_size, t == 1)
                 # offsetting since the returned smallest_k() indices were slice-relative
-                best_hyp_indices_np[rows] += rows.start
+                best_hyp_indices[rows] += rows.start
 
-            # convert back to mx.ndarray again
-            best_hyp_indices[:] = best_hyp_indices_np
-            best_word_indices[:] = best_word_indices_np
-            scores_accumulated[:] = np.expand_dims(scores_accumulated_np, axis=1)
```
Comment (mjpost) [inline sockeye/inference.py]: It is nice to see that go.

File: sockeye/arguments.py [hunk 4]
```diff
@@ -767,6 +767,15 @@ def add_training_args(params):
                               help="The MXNet kvstore to use. 'device' is recommended for single process training. "
                                    "Use any of 'dist_sync', 'dist_device_sync' and 'dist_async' for distributed "
                                    "training. Default: %(default)s.")
+    train_params.add_argument("--gradient-compression-type",
+                              type=str,
+                              default=None,
+                              choices=[None, "2bit"],
```
Comment (davvil) [inline sockeye/arguments.py]: Add "2bit" as a constant?

Comment (fhieber) [reply sockeye/arguments.py]: done

File: CHANGELOG.md [hunk 5]
```diff
@@ -10,6 +10,16 @@ Note that Sockeye has checks in place to not translate with an old model that wa
 
 Each version section may have have subsections for: _Added_, _Changed_, _Removed_, _Deprecated_, and _Fixed_.
 
+## [1.16.0]
+### Changed
+- Update to [MXNet 1.0.0](https://github.com/apache/incubator-mxnet/tree/1.0.0) which adds more advanced indexing
+features, benefitting the beam search implementation.
+- `--kvstore` now accepts 'nccl' value. Only works if MXNet was compiled with `USE_NCCL=1`
```
Comment (tdomhan) [inline CHANGELOG.md]: period missing ;)
